### PR TITLE
upgraded coinmaster

### PR DESCRIPTION
**Pull request by matpb**
## Meta  

Notion task: https://www.notion.so/92bddda861cd41a38ec8c2aa9c5d69d0?v=116d739390814b2e97702e668fcc601b&p=d8d828fb6b454b5fbecf229553c11b18&pm=s

Closing #12 and #13 

## Changelog  
- new wasm bindings supporting cosmwasm_std standard

> This is API breaking for smart contract developers. See [Coinmaster README](https://github.com/noria-net/noria/blob/a442210ba6831b79cf3a458138228b485f210f7a/x/coinmaster/README.md)  

- new coinmaster params unit test for `SetParams` validation
- renamed coinmaster messages to avoid message collision check in Cosmos SDK 47

> MsgMint -> MsgCoinmasterMint
> MsgBurn -> MsgCoinmasterBurn

- Coinmaster now supports multiple whitelisted minters, including support for smart contracts.

## Tests

- v1.0.0 -> this PR is green
- All `/x/coinmaster` tests passing green

## Tag

I believe this PR can be merged with main and be tagged for the next `v1.1.0` release.

### CODE REVIEW

File: x/coinmaster/bindings/message_plugin.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		}
+	}
+	return m.wrapped.DispatchMsg(ctx, contractAddr, contractIBCPortID, msg)
+}
+
+// mint
+func (m *CustomMessenger) mint(ctx sdk.Context, contractAddr sdk.AccAddress, msg *coinmastertypes.MsgCustomCoinmasterMint) ([]sdk.Event, [][]byte, error) {
+	bz, err := PerformMint(m.coinmaster, ctx, contractAddr, msg)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrap(err, "perform mint denom")
```
Comment (jumpdest7d) [inline x/coinmaster/bindings/message_plugin.go:61]: can the error message be expanded? Might be a bit hard to follow this error message.

File: x/coinmaster/bindings/query_plugin.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+// CustomQuerier dispatches custom CosmWasm bindings queries.
+func CustomQuerier(qp *QueryPlugin) func(ctx sdk.Context, request json.RawMessage) ([]byte, error) {
+	return func(ctx sdk.Context, request json.RawMessage) ([]byte, error) {
+		var contractQuery coinmastertypes.CoinmasterQuery
+		if err := json.Unmarshal(request, &contractQuery); err != nil {
+			return nil, sdkerrors.Wrap(err, "coinmaster query")
+		}
+		if contractQuery.Coinmaster == nil {
+			return nil, sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "nil coinmaster field")
```
Comment (jumpdest7d) [inline x/coinmaster/bindings/query_plugin.go:22]: This differs from the guard check done on [message_plugin.go:42](https://github.com/noria-net/noria/pull/14/files#diff-cf62c38355db48c6113b10f39ae49bedba7e9a8e7b0739b85c5633d6c1caee23R42) - is that expected?

File: x/coinmaster/keeper/params.go [hunk 3]
```diff
@@ -14,8 +18,34 @@ func (k Keeper) GetParams(ctx sdk.Context) types.Params {
 }
 
 // SetParams set the params
-func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
+func (k Keeper) SetParams(ctx sdk.Context, params types.Params) (err error) {
+	if params.Minters != "" {
+		// validate that params.Minters is a comma separated list of addresses
+		mintersRegex := (`^([a-z0-9]{40,80},)*[a-z0-9]{40,80}$`)
```
Comment (jumpdest7d) [inline x/coinmaster/keeper/params.go:24]: Let's throw the regex into a utility function, something like "splitAddresses" and then we can surround that with a bunch of edge-case and happy case tests.

I also wonder if there's a utility function out there that splits comma-separated addresses for us already?

**Review by jumpdest7d — DISMISSED** [review]
Yo Mat! Approved with a few comments. Nothing critical.

File: x/coinmaster/bindings/message_plugin.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+			return m.mint(ctx, contractAddr, coinmasterMsg.Mint)
+		}
+		if coinmasterMsg.Burn != nil {
+			return m.burn(ctx, contractAddr, coinmasterMsg.Burn)
+		}
+	}
+	return m.wrapped.DispatchMsg(ctx, contractAddr, contractIBCPortID, msg)
+}
+
+// mint
```
[the remark this answers is not in the document]
Comment (matpb) [reply x/coinmaster/bindings/message_plugin.go:57]: That is actually crucial for other messages to be dispatched by the global query plugin. Coinmaster only intercepts `CustomMsg` messages that contains the `coinmaster` decorator.

File: x/coinmaster/bindings/message_plugin.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+var _ wasmkeeper.Messenger = (*CustomMessenger)(nil)
+
+// DispatchMsg executes on the contractMsg.
+func (m *CustomMessenger) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddress, contractIBCPortID string, msg wasmvmtypes.CosmosMsg) ([]sdk.Event, [][]byte, error) {
+	if msg.Custom != nil {
+		// only handle the happy path where this is really creating / minting / swapping ...
+		// leave everything else for the wrapped version
+		var contractMsg coinmastertypes.CoinmasterMsg
+		if err := json.Unmarshal(msg.Custom, &contractMsg); err != nil {
+			return nil, nil, sdkerrors.Wrap(err, "coinmaster msg")
```
[the remark this answers is not in the document]
Comment (matpb) [reply x/coinmaster/bindings/message_plugin.go]: Great idea, thanks! Added better error handling and and error bubbling in 09f84f9a0506252b21569b18c79ecf7d7dea67ea

File: x/coinmaster/bindings/message_plugin.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+		}
+		if contractMsg.Coinmaster == nil {
+			// return nil, nil, sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "nil coinmaster field")
```
[the remark this answers is not in the document]
Comment (matpb) [reply x/coinmaster/bindings/message_plugin.go]: Good catch, removed in 09f84f9a0506252b21569b18c79ecf7d7dea67ea

File: x/coinmaster/bindings/message_plugin.go (hunk 1 above)
Comment (matpb) [reply x/coinmaster/bindings/message_plugin.go]: Absolutely right, improved in 09f84f9a0506252b21569b18c79ecf7d7dea67ea

File: x/coinmaster/bindings/query_plugin.go (hunk 2 above)
Comment (matpb) [reply x/coinmaster/bindings/query_plugin.go:21]: The guards are ok because queries are different than messages and do not get dispatched sequencially like messages do. But that comment did make me realize some other issues with the query plugin I wrote (I was today years old when I learned the importance of not setting `omitempty` for wasm bindings!). Fixed in 5c1cd3643ff6c436b6fa4dd33091d8e169668982

File: x/coinmaster/keeper/params.go (hunk 3 above)
Comment (matpb) [reply x/coinmaster/keeper/params.go]: Good idea: refactored in 825a400a26d021b5300c52e382ddab29babcd231

**Comment by matpb** [comment]
All comments were considered, improvements added. 1 comment was clarified and didn't need modification. Review required, @jumpdest7d

File: x/coinmaster/bindings/query_plugin.go (hunk 2 above)
Comment (jumpdest7d) [reply x/coinmaster/bindings/query_plugin.go:21]: I too was today years old when I learned what `omitempty` does!

**Review by jumpdest7d — APPROVED** [review]
LGTM @matpb!

File: x/coinmaster/bindings/message_plugin.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jumpdest7d) [reply x/coinmaster/bindings/message_plugin.go:56]: That's awesome to know, thank you!
